### ISSUE TITLE
Enhancement: [no-redundant-type-constituents] Use assignability checking for redundancy checks

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#7742]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-redundant-type-constituents

### Description

Right now, `no-redundant-type-constituents` is very limited:

> This rule plays it safe and only works with bottom types, top types, and comparing literal types to primitive types.

But sometimes an intersection or union type will have two constituents where one constituent is a superset of the other. It'd be nice to report on those cases too. Or, heck, even cases where they're deeply equal, but declared as different object types.

### Fail

```typescript
type ShallowEqual = number | number;
type DeepEqual = { a: string } | { a: string };
type Superset = { a: string } | { a: "b" | "c" };
```

### Pass

```typescript
type ShallowEqual = string | number;
type DeepEqual = { a: string } | { a: number };
type Superset = { a: string } | { a: "b" | "c" | 0 };
```

### Additional Info

If we could use a type assignability / relationship API (https://github.com/microsoft/TypeScript/issues/9879), then we could catch those. Filing this long-standing rule feature request for tracking.

**Comment by JoshuaKGoldberg** [comment]
Oh hey this should be unblocked now per #7936! 🚀

**Comment by JoshuaKGoldberg** [comment]
#10744 got the job done as described, but there are a lot more ramifications from this change than we realized. We reverted it just now and are re-opening this issue.

* #11787
* #11790
* #11800
* #11810
* #11811 

We're definitely still interested in using assignability checking for no-redundant-type-constituents! Just, a PR to add that feature will need to address those issues.

**Comment by Tezra** [comment]
Before reopening this, should decide how this rule should handle "strictNullChecks". When false, simple assignability says `string | undefined` is redundant, but when compiling the `| undefined` still gets emitted, which affects consuming projects (that may be strict) making it not redundant in the wider picture. Both ways are arguably a bug unless documented.

Because of the emit thing, I would expect this rule to always treat types as if strictNullChecks where true. That would probably require a change to the TypeScript assignability API to allow overriding this and anything else that would affect assignability. Which would make this much more difficult to implement properly. (Could do an exception for null / undefined. I don't think there are currently any other config options that affect assignability)

**Comment by bradzacher** [comment]
The issue is that it's really hard for us to support `strictNullChecks` in rules because `null | undefined` is omitted by TS in the local type information -- as in when we use their APIs to query for the type information `null | undefined` is literally not there.

That's why some of our rules are literally setup to crash if you have `strictNullChecks: false`.

`strictNullChecks: true` has been the recommendation since its release, and with the upcoming semver major release of TS they're going to be setting `strictNullChecks: true` as the default for all projects that don't explicitly set it to `false`.

Really IMO we should also be heading towards the setup where using `strictNullChecks: false` is officially unsupported and "undefined behaviour" for all of our rules.

**Comment by Tezra** [comment]
In that case it may be better to just update the rule description with a note that if you are using `strictNullChecks: false`, `null` and `undefined` will always be redundant, and if using both ` | undefined` and `strictNullChecks: false` in your project, you may want to disable this rule.

If there is a way to get it to play nice with null / undefined though as if strictNullChecks where true though, that would be helpful for projects that are in the process of updating code to strict.

**Comment by kirkwaiblinger** [comment]
I'm in favor of just documenting `strictNullChecks` as a limitation/"when not to use it".

**Comment by ArnaudBarre** [comment]
Hello the TS-ESLint team!
For info when I ported this rule to my TS linter, I did use the newly available [isTypeAssignableTo](https://github.com/ArnaudBarre/tsl/blob/caa8a2e83d9de68a7dc1caa172152d17e53677af/src/rules/noRedundantTypeConstituents/noRedundantTypeConstituents.ts#L80). It's not used a lot, so I can't guaranty that the implementation doesn't have edge cases, but here is how it behaves on the reports above:
- #11787 No crash
- #11790 Does report also, and feels legit IMO
- #11800 I really don't care about people wanting to use a linter with `strictNullChecks: false`
- #11810 Does't report valid case 1., report cases 2&3 because object does include arrays, blob, Error, ...
- #11811 Doesn't report

Feel free to take part of the implementation that could be useful to you, I did not read https://github.com/typescript-eslint/typescript-eslint/pull/10744 so I don't know what are the differences
